lib/synapse/service_watcher/base.rb

```diff
@@ -95,13 +100,44 @@ def validate_discovery_opts
     end
 
     def set_backends(new_backends)
-      if @keep_default_servers
-        @backends = @default_servers + new_backends
+      # Aggregate and deduplicate all potential backend service instances.
+      new_backends = (new_backends + (@keep_default_servers ? @default_servers : [])).uniq {|b|
```

Comment:
oh i see. i think this would read a little cleaner if it was not all on one line. i would say:
```ruby
new_backends += @default_servers if @keep_default_servers
new_backends.uniq!{|b|  [b['host'], b['port'], b.fetch('name', '')]}
```